src/backend/binaryen/wasm_expr_gen.ts

```diff
@@ -2795,133 +2839,144 @@ export class WASMExpressionGen {
+            ],
+            binaryen.anyref,
+        );
+    }
+
+    private dynGetInfcUnionPropertyHelper(
+        objRef: binaryen.ExpressionRef,
+        flagAndIndexRef: binaryen.ExpressionRef,
+        valueType: ValueType,
+        propTypeIdRef: binaryen.ExpressionRef,
+    ) {
+        const indexRef = FunctionalFuncs.getPropIndexFromObj(
+            this.module,
+            flagAndIndexRef,
+        );
+        const flagRef = FunctionalFuncs.getPropFlagFromObj(
+            this.module,
+            flagAndIndexRef,
+        );
+
+        let condition: binaryen.ExpressionRef;
+        const valueTypeId = FunctionalFuncs.getPredefinedTypeId(valueType);
+        if (valueTypeId >= PredefinedTypeId.CUSTOM_TYPE_BEGIN) {
+            condition = FunctionalFuncs.isPropTypeIdIsObject(
+                this.module,
+                propTypeIdRef,
             );
-            binaryenCAPI._BinaryenIfSetIfFalse(curIfExpr, branchForOpt);
         } else {
-            binaryenCAPI._BinaryenIfSetIfFalse(
-                curIfExpr,
-                this.module.unreachable(),
+            condition = FunctionalFuncs.isPropTypeIdCompatibal(
+                this.module,
+                this.module.i32.const(valueTypeId),
+                propTypeIdRef,
             );
         }
-
-        return this.module.if(
-            FunctionalFuncs.isPropertyExist(this.module, flagAndIndexRef),
-            FunctionalFuncs.generateDynUndefined(this.module),
-            ifExpr,
+        const realValueRef = this.getIndirectValueRef(
+            objRef,
+            indexRef,
+            flagRef,
+            valueType,
+        );
+        const anyTypedRealValueRef = FunctionalFuncs.boxNonLiteralToAny(
+            this.module,
+            realValueRef,
+            valueType.kind,
         );
+        return this.module.if(condition, anyTypedRealValueRef);
     }
 
-    private dynGetInfcUnionPropertyHelper(
-        ref: binaryen.ExpressionRef,
-        flagAndIndexRef: binaryen.ExpressionRef,
+    setIndirectValueRef(
+        objRef: binaryen.ExpressionRef,
+        indexRef: binaryen.ExpressionRef,
+        flagRef: binaryen.ExpressionRef,
         valueType: ValueType,
-        typeIdRef: binaryen.ExpressionRef,
+        valueRef: binaryen.ExpressionRef,
     ) {
-        const indexRef = this.getPropIndexFromObj(flagAndIndexRef);
-        let cond: binaryen.ExpressionRef;
-        let getRealValueRef: binaryen.ExpressionRef;
-        const kind = valueType.kind;
-        /** Seems it only used when creating closure for method */
-        const wasmType = this.wasmTypeGen.getWASMType(valueType);
-
-        switch (kind) {
+        const propValueTypeRef = this.wasmTypeGen.getWASMValueType(valueType);
+        const propTypeRef = this.wasmTypeGen.getWASMType(valueType);
+        let res: binaryen.ExpressionRef;
+
+        /* just for providing get_indirect path, not used in real custom envionment */
```

Comment:
what does this comment mean?

Reply:
Our struct_indirect APIs also including `i64` and `f32` version, but in real custom codes, `xx.a` will always be a valid type in ts, so `i64` and `f32` will never occured, I just leave these here to show `i64` and `f32` can be getted here, but I never see on sample go through this.